lib/ansible/cli/doc.py

```diff
@@ -71,7 +73,197 @@ class PluginNotFound(Exception):
+
+        :param collection_filter: A string containing the FQCN of a collection which will be
+            used to limit results. This filter will take precedence over the name_filters.
+
+        :returns: A set of tuples consisting of: role name, collection name, collection path
+        """
+        found = set()
+        b_colldirs = list_collection_dirs(coll_filter=collection_filter)
+        for b_path in b_colldirs:
+            path = to_text(b_path, errors='surrogate_or_strict')
+            collname = _get_collection_name_from_path(b_path)
+
+            roles_dir = os.path.join(path, 'roles')
+            if os.path.exists(roles_dir):
+                for entry in os.listdir(roles_dir):
+                    full_path = os.path.join(roles_dir, entry, 'meta', self.ROLE_ARGSPEC_FILE)
+                    if os.path.exists(full_path):
+                        if name_filters is None:
+                            found.add((entry, collname, path))
+                        else:
+                            # Name filters might contain a collection FQCN or not.
+                            for fqcn in name_filters:
+                                if len(fqcn.split('.')) == 3:
+                                    (ns, col, role) = fqcn.split('.')
+                                    if '.'.join([ns, col]) == collname and entry == role:
+                                        found.add((entry, collname, path))
+                                elif fqcn == entry:
+                                    found.add((entry, collname, path))
+        return found
+
+    def _create_role_list(self, roles_path, collection_filter=None):
+        """Return a dict describing the listing of all roles with arg specs.
+
+        :param role_paths: A tuple of one or more role paths.
+
+        :returns: A dict indexed by role name, with 'collection' and 'entry_points' keys per role.
+
+        Example return:
+
+            results = {
+               'roleA': {
+                  'collection': '',
+                  'entry_points': {
+                     'main': 'Short description for main'
+                  }
+               },
+               'a.b.c.roleB': {
+                  'collection': 'a.b.c',
+                  'entry_points': {
+                     'main': 'Short description for main',
+                     'alternate': 'Short description for alternate entry point'
+                  }
+               'x.y.z.roleB': {
+                  'collection': 'x.y.z',
+                  'entry_points': {
+                     'main': 'Short description for main',
+                  }
+               },
+            }
+
+        """
+        if not collection_filter:
+            roles = self._find_all_normal_roles(roles_path)
+        else:
+            roles = []
+        collroles = self._find_all_collection_roles(collection_filter=collection_filter)
+
+        result = {}
+
+        def build_summary(role, collection, argspec):
```

Comment:
Does this need to be a nested function? The overall cyclomatic complexity of this method is 8. I bet it could be reduced by removing this nested function. It would also make this method (and the function) easier to test.

Reply:
TIL that "cyclomatic complexity" is a thing.  :)

Couple of questions on this:

1) Are you suggesting to just duplicate the code from `build_summary()` within each `for` loop? TBH, I would hate to do that (and was the reason for the nested function).

2) I don't see how having the nested function within `_create_role_list()` makes testing harder, but that's probably because I'm unfamiliar with cyclomatic complexity. Could you elaborate?

Reply:
FYI, I'm moving the nest func out to a private class method.

Reply:
> Are you suggesting to just duplicate the code from build_summary() within each for loop? 

Nope. It just seems like this nested function could be a method or a standalone function.

> I don't see how having the nested function within _create_role_list() makes testing harder,

In order to unit test `build_summary()` you would need an instance of `RoleMixin`, then call `_create_role_list()`, and then write the test in such as way as to only run `build_summary()`. It just makes it harder to get to _just_ `build_summary()` for testing, which makes it harder to throw tons of varying inputs at `build_summary()`.

Reply:
> FYI, I'm moving the nest func out to a private class method.

Awesome! The complixity of `_create_role_list()` is now 4. That's what winning looks like. 😎